### PR TITLE
Implement Block interface

### PR DESCRIPTION
**Pull request by firzanarmani**
Let me know if there's anything else I should consider. Based on proposal at #8 

When given the approval, I'll make a PR for the actual Block, Title and Body classes.

BlockManager still needs work to tie up with the ObservableBlockTree!

### CODE REVIEW

**Comment by kevinputera** [comment]
Hey @firzanarmani ! I think you should create the skeleton classes for `Title` and `Body` alongside this PR, to make sure the code builds!

File: src/main/java/com/notably/model/block/BlockModel.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    Title getTitle();
+
+    /**
+     * Gets the body content of a block.
+     */
+    Body getBody();
+
+    /**
+     * Gets the parent block of a block, if possible.
+     * Returns an {@code Optional<Empty>} if the block is the root block.
```
Comment (kevinputera) [inline src/main/java/com/notably/model/block/BlockModel.java]: Nit: I believe `Optional<Empty>` should be `Optional.empty()` instead?

File: src/main/java/com/notably/model/block/BlockModel.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Finds a child block of a block, with the given input.
+     */
+    Optional<Block> getChild(Title title);
+
+    /**
+     * Replaces a child block of a block, with a new child block.
+     */
+    void setChild(Block oldBlock, Block newBlock);
```
Comment (kevinputera) [inline src/main/java/com/notably/model/block/BlockModel.java]: Should this `Block oldBlock` parameter instead be `Title title`?
If we take a look at `java.util.List`'s [set](https://docs.oracle.com/javase/8/docs/api/java/util/List.html) method, we can see that the parameters are `(int index, T item)`. Should we follow something similar here?

File: src/main/java/com/notably/model/block/BlockModel.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Removes a specified child block from a block.
+     * @param toRemove
+     */
+    void removeChild(Block toRemove);
+
+    /**
+     * Checks if a block is a root block.
+     */
+    boolean isRoot();
```
Comment (kevinputera) [inline src/main/java/com/notably/model/block/BlockModel.java]: I'm curious! How do you imagine this `isRoot` method be used (in what scenario)?

File: src/main/java/com/notably/model/block/BlockModel.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package com.notably.model.block;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * API of the BlockModel component.
+ */
+public interface BlockModel {
```
Comment (kevinputera) [inline src/main/java/com/notably/model/block/BlockModel.java]: Is this supposed to be `public interface Block`?

**Review by kevinputera — COMMENTED** [review]
Just a few comments/questions. Otherwise, looks great!

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 4 above)
Comment (firzanarmani) [reply src/main/java/com/notably/model/block/BlockModel.java]: I intended to have the implementing class be `Block`. What do you think?

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 1 above)
Comment (firzanarmani) [reply src/main/java/com/notably/model/block/BlockModel.java]: Will fix! Thanks for spotting my mistake!

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 2 above)
Comment (firzanarmani) [reply src/main/java/com/notably/model/block/BlockModel.java]: Hmm, makes sense. I'll make the change

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 4 above)
Comment (kevinputera) [reply src/main/java/com/notably/model/block/BlockModel.java]: Ah, how about calling the interface `Block`, then the concrete class `BlockImpl`? I believe @johannagwan  is using this naming convention too for `SuggestedItem` and `SuggestedItemImpl`. What do you think?

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 1 above)
Comment (kevinputera) [reply src/main/java/com/notably/model/block/BlockModel.java]: Awesome!

File: src/main/java/com/notably/model/block/Title.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+public class Title {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Titles should only contain alphanumeric characters and spaces, and it should not be blank";
+
+    /*
+     * The first character of the title must not be a whitespace,
+     * otherwise " " (a blank string) becomes a valid input.
+     */
+    public static final String VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
```
Comment (kevinputera) [inline src/main/java/com/notably/model/block/Title.java]: This regex can be rewritten as `"^\\p{Alnum}[\\p{Alnum} ]*$"`.
The `[]` in the first `\\p{Alnum}` isn't really necessary, as it is only a single item.
The `^$` symbol anchors the regex to the start and end of the string (whole string)

Comment (kevinputera) [reply src/main/java/com/notably/model/block/Title.java]: Actually, after reading more about how `String#matches` work, I think the `^$` part is not necessary

**Review by kevinputera — COMMENTED** [review]
Some more comments!

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 2 above)
Comment (kevinputera) [reply src/main/java/com/notably/model/block/BlockModel.java]: Awesome!

**Comment by johannagwan** [comment]
LGTM!

**Comment by firzanarmani** [comment]
>Ah, how about calling the interface Block, then the concrete class BlockImpl? I believe @johannagwan is using this naming convention too for SuggestedItem and SuggestedItemImpl. What do you think?

Fine with me! I'll change accordingly!

File: src/main/java/com/notably/model/block/Title.java (hunk 5 above)
Comment (firzanarmani) [reply src/main/java/com/notably/model/block/Title.java]: Changed! :)

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 3 above)
Comment (firzanarmani) [reply src/main/java/com/notably/model/block/BlockModel.java]: Root block will also be a regular block but with
- `title = "root"`
- `body = null`
- `parent = null`
- `children` as per normal block behaviour

So isRoot() will be useful later by even others, to check if the block is the root block. It's also written explicitly as a method for easier identification.

File: src/main/java/com/notably/model/block/BlockModel.java (hunk 4 above)
Comment (firzanarmani) [reply src/main/java/com/notably/model/block/BlockModel.java]: Fine with me! I'll change to match the naming convention!

**Comment by HemanshuGandhi** [comment]
LGTM!

**Review by HemanshuGandhi — APPROVED** [review]

**Review by kevinputera — APPROVED** [review]
LGTM!

**Pull request merged**